lib/libtsd/tsd_hash.3
.\"-
.\" Copyright (c) 2015 The University of Oslo
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior written
.\"    permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd February 5, 2015
.Dt TSD_HASH 3
.Os
.Sh NAME
.Nm tsd_hash ,
.Nm tsd_strhash
.Nd hash functions
.Sh LIBRARY
.Lb libtsd
.Sh SYNOPSIS
.In stdint.h
.In tsd/hash.h
.Ft uint8_t
.Fn tsd_hash "const void *data" "size_t len"
.Ft uint8_t
.Fn tsd_strhash "const char *str"
.Sh DESCRIPTION
The
.Fn tsd_hash
function returns an 8-bit hash of the
.Va len
first bytes of the object pointed to by
.Va data .
.Pp
The
.Fn tsd_strhash
function returns an 8-bit hash of the NUL-terminated string pointed to
by
.Va str ,
not including the terminating NUL.
It is equivalent to calling
.Li tsd_hash(str, strlen(str)) .
.Sh SEE ALSO
.Xr tsd_sha1 3
.Sh REFERENCES
.Rs
.%A Pearson, Peter K.
.%D June 1990
.%T Fast Hashing of Variable-Length Text Strings
.%J Communications of the ACM
.%V 33
.%N 6
.%P 677
.%O doi:10.1145/78973.78978
.Re
.Sh AUTHORS
The
.Fn tsd_hash
functions and this manual page were written for the University of Oslo by
.An Dag-Erling Sm\(/orgrav Aq Mt d.e.smorgrav@usit.uio.no .


lib/libtsd/tsd_hash.c
/*-
 * Copyright (c) 2014 The University of Oslo
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior written
 *    permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <stdint.h>
#include <string.h>

#include <tsd/hash.h>

/*
 * Implementation of the hash function described in Pearson, Peter
 * K. (June 1990), "Fast Hashing of Variable-Length Text Strings",
 * Communications of the ACM 33 (6): 677, doi:10.1145/78973.78978
 */

static const uint8_t T[256] = {
	0x01,	0x57,	0x31,	0x0c,	0xb0,	0xb2,	0x66,	0xa6,
	0x79,	0xc1,	0x06,	0x54,	0xf9,	0xe6,	0x2c,	0xa3,
	0x0e,	0xc5,	0xd5,	0xb5,	0xa1,	0x55,	0xda,	0x50,
	0x40,	0xef,	0x18,	0xe2,	0xec,	0x8e,	0x26,	0xc8,
	0x6e,	0xb1,	0x68,	0x67,	0x8d,	0xfd,	0xff,	0x32,
	0x4d,	0x65,	0x51,	0x12,	0x2d,	0x60,	0x1f,	0xde,
	0x19,	0x6b,	0xbe,	0x46,	0x56,	0xed,	0xf0,	0x22,
	0x48,	0xf2,	0x14,	0xd6,	0xf4,	0xe3,	0x95,	0xeb,
	0x61,	0xea,	0x39,	0x16,	0x3c,	0xfa,	0x52,	0xaf,
	0xd0,	0x05,	0x7f,	0xc7,	0x6f,	0x3e,	0x87,	0xf8,
	0xae,	0xa9,	0xd3,	0x3a,	0x42,	0x9a,	0x6a,	0xc3,
	0xf5,	0xab,	0x11,	0xbb,	0xb6,	0xb3,	0x00,	0xf3,
	0x84,	0x38,	0x94,	0x4b,	0x80,	0x85,	0x9e,	0x64,
	0x82,	0x7e,	0x5b,	0x0d,	0x99,	0xf6,	0xd8,	0xdb,
	0x77,	0x44,	0xdf,	0x4e,	0x53,	0x58,	0xc9,	0x63,
	0x7a,	0x0b,	0x5c,	0x20,	0x88,	0x72,	0x34,	0x0a,
	0x8a,	0x1e,	0x30,	0xb7,	0x9c,	0x23,	0x3d,	0x1a,
	0x8f,	0x4a,	0xfb,	0x5e,	0x81,	0xa2,	0x3f,	0x98,
	0xaa,	0x07,	0x73,	0xa7,	0xf1,	0xce,	0x03,	0x96,
	0x37,	0x3b,	0x97,	0xdc,	0x5a,	0x35,	0x17,	0x83,
	0x7d,	0xad,	0x0f,	0xee,	0x4f,	0x5f,	0x59,	0x10,
	0x69,	0x89,	0xe1,	0xe0,	0xd9,	0xa0,	0x25,	0x7b,
	0x76,	0x49,	0x02,	0x9d,	0x2e,	0x74,	0x09,	0x91,
	0x86,	0xe4,	0xcf,	0xd4,	0xca,	0xd7,	0x45,	0xe5,
	0x1b,	0xbc,	0x43,	0x7c,	0xa8,	0xfc,	0x2a,	0x04,
	0x1d,	0x6c,	0x15,	0xf7,	0x13,	0xcd,	0x27,	0xcb,
	0xe9,	0x28,	0xba,	0x93,	0xc6,	0xc0,	0x9b,	0x21,
	0xa4,	0xbf,	0x62,	0xcc,	0xa5,	0xb4,	0x75,	0x4c,
	0x8c,	0x24,	0xd2,	0xac,	0x29,	0x36,	0x9f,	0x08,
	0xb9,	0xe8,	0x71,	0xc4,	0xe7,	0x2f,	0x92,	0x78,
	0x33,	0x41,	0x1c,	0x90,	0xfe,	0xdd,	0x5d,	0xbd,
	0xc2,	0x8b,	0x70,	0x2b,	0x47,	0x6d,	0xb8,	0xd1,
};

uint8_t
tsd_hash(const void *data, size_t len)
{
	const uint8_t *p;
	uint8_t h;

	for (p = data, h = 0; len > 0; ++p, --len)
		h = T[h ^ *p];
	return (h);
}

uint8_t
tsd_strhash(const char *str)
{
	uint8_t h;

	for (h = 0; *str != '\0'; ++str)
		h = T[h ^ (uint8_t)*str];
	return (h);
}
